### PR TITLE
Fix and disable encoding for nanosecond statistics in ORC writer

### PR DESCRIPTION
**Pull request by vuule**
## Description
Issue https://github.com/rapidsai/cudf/issues/14325

Use uint when reading/writing nano stats because nanoseconds have int32 encoding (different from both unit32 and sint32, _obviously_), which does not use zigzag. 
sint32 uses zigzag, and unit32 does not allow negative numbers, so we can use uint since we'll never have negative nanoseconds.

Also disabled the nanoseconds because it should only be written after ORC-135; we don't write the version so readers get confused if nanoseconds are there. Planning to re-enable once we start writing the version.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/src/io/orc/orc.hpp [hunk 1]
```diff
@@ -41,6 +41,9 @@ static constexpr uint32_t block_header_size = 3;
 // Seconds from January 1st, 1970 to January 1st, 2015
 static constexpr int64_t orc_utc_epoch = 1420070400;
 
+static constexpr int32_t DEFAULT_MIN_NANOS = 0;
+static constexpr int32_t DEFAULT_MAX_NANOS = 999999;
```
Comment (vyasr) [inline cpp/src/io/orc/orc.hpp:45]: Could we document why these are the min and max values? These aren't powers of 2 so I assume it's not integer size based, is it limited by some spec?

File: cpp/src/io/orc/stats_enc.cu [hunk 2]
```diff
@@ -27,6 +27,8 @@ namespace cudf::io::orc::gpu {
 
 using strings::detail::fixed_point_string_size;
 
+constexpr bool enable_nanosecond_statistics = false;
```
Comment (vyasr) [inline cpp/src/io/orc/stats_enc.cu:30]: Could we link https://github.com/rapidsai/cudf/issues/14325 here as this flag is what needs to be be flipped to add the currently excluded statistic?

**Review by vyasr — APPROVED** [review]
A couple of requests for comments, otherwise the change looks fine insofar as it implements exactly what the description promises. Based on the discussion in #14325 it looks like this fix also stops the crash, which is the primary goal for 23.12

File: cpp/src/io/orc/orc.cpp [hunk 3]
```diff
@@ -182,6 +182,9 @@ void ProtobufReader::read(timestamp_statistics& s, size_t maxlen)
                        field_reader(5, s.minimum_nanos),
                        field_reader(6, s.maximum_nanos));
   function_builder(s, maxlen, op);
+  // adjust nanoseconds because someone had a bright idea to store nanoseconds + 1 in the protobuf
+  if (s.minimum_nanos.has_value()) { --s.minimum_nanos.value(); }
+  if (s.maximum_nanos.has_value()) { --s.maximum_nanos.value(); }
```
Comment (ttnghia) [inline cpp/src/io/orc/orc.cpp:187]: Since we are using unsigned int, will this underflow?

File: cpp/src/io/orc/orc.hpp (hunk 1 above)
Comment (ttnghia) [reply cpp/src/io/orc/orc.hpp:45]: Why don't we upgrade the type of these constants to `uint32_t`?

File: cpp/src/io/orc/orc.cpp (hunk 3 above)
Comment (vuule) [reply cpp/src/io/orc/orc.cpp:187]: The nanoseconds are stored as `value+1`, so zero is not a valid content here. We found out the hard way about this :D

Comment (ttnghia) [reply cpp/src/io/orc/orc.cpp:187]: Can you add a comment to clarify that please? So we won't forget about it.

File: cpp/src/io/orc/orc.hpp (hunk 1 above)
Comment (vuule) [reply cpp/src/io/orc/orc.hpp:45]: These do come from specs.
Nanosecond timestamp statistics are stored at millisecond precision + the leftover nanoseconds. This is why the max is 999999, one more and you have a full millisecond.

I can derive the max from `chrono`. Let me know.

Should we use unsigned here? AFAIK we should avoid any uints for any arithmetic.

Comment (vyasr) [reply cpp/src/io/orc/orc.hpp:45]: I'm fine using the signed ints, and I don't feel the need to use chrono. Just document that we store as ms + ns so 1e6 ns is the largest you'll ever need.

Comment (ttnghia) [reply cpp/src/io/orc/orc.hpp:45]: But will we have issue with sign vs unsign comparison?

Comment (vuule) [reply cpp/src/io/orc/orc.hpp:45]: The only place these are used is to compare against `min_ns_remainder`/`max_ns_remainder`, and these are signed.

Comment (ttnghia) [reply cpp/src/io/orc/orc.hpp:45]: Great.

**Review by ttnghia — APPROVED** [review]

File: cpp/src/io/orc/orc.cpp (hunk 3 above)
Comment (ttnghia) [reply cpp/src/io/orc/orc.cpp:187]: Oh, approved but forgot about this :smile:

Comment (vuule) [reply cpp/src/io/orc/orc.cpp:187]: no worries, I haven't ;)

**Pull request merged**